userns.go

```diff
@@ -0,0 +1,220 @@
+package storage
+
+import (
+	"os"
+	"os/user"
+	"strconv"
+
+	"github.com/containers/common/pkg/unshare"
+	"github.com/containers/storage/pkg/idtools"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func getAutoUserNsMappings(username string) ([]idtools.IDMap, []idtools.IDMap, error) {
+	var uids, gids []idtools.IDMap
+
+	if unshare.IsRootless() {
+		username = os.Getenv("USER")
+		if username == "" {
+			var id string
+			if os.Geteuid() == 0 {
+				id = strconv.Itoa(unshare.GetRootlessUID())
+			} else {
+				id = strconv.Itoa(os.Geteuid())
+			}
+			userID, err := user.LookupId(id)
+			if err == nil {
+				username = userID.Username
+			}
+		}
+	} else {
+		if username == "" {
+			username = "storage"
+		}
+	}
+	mappings, err := idtools.NewIDMappings(username, username)
+	if err != nil {
+		logrus.Errorf("cannot find mappings for user %q: %v", username, err)
+	} else {
+		uids = mappings.UIDs()
+		gids = mappings.GIDs()
+	}
+	return uids, gids, nil
+}
+
+func (s *store) getAvailableMappings() ([]idtools.IDMap, []idtools.IDMap, error) {
+	uids := s.autoUIDMap
+	gids := s.autoGIDMap
+
+	if !unshare.IsRootless() {
+		// No mapping to inner namespace needed
+		return copyIDMap(uids), copyIDMap(gids), nil
+	}
+
+	// We are already inside on the rootless user namespace.
+	// So we need to remap the configured mappings to what is available inside of the userns.
+	totaluid := 0
+	totalgid := 0
+	for _, u := range uids {
+		totaluid += u.Size
+	}
+	for _, g := range gids {
+		totalgid += g.Size
+	}
+
+	u := []idtools.IDMap{{ContainerID: 0, HostID: 1, Size: totaluid}}
+	g := []idtools.IDMap{{ContainerID: 0, HostID: 1, Size: totalgid}}
+	return u, g, nil
+}
+
+// getAutoUserNS creates an automatic user namespace
+func (s *store) getAutoUserNS(size int, imageName string) ([]idtools.IDMap, []idtools.IDMap, error) {
+	containers, err := s.Containers()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	availableUIDs, availableGIDs, err := s.getAvailableMappings()
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "cannot read mappings")
+	}
+
+	var usedUIDs, usedGIDs []idtools.IDMap
+	for _, c := range containers {
+		usedUIDs = append(usedUIDs, c.UIDMap...)
+		usedGIDs = append(usedGIDs, c.GIDMap...)
+	}
+
+	if size <= 0 {
+		size = 4096
```

Comment:
Should this be defined in containers/storage?

Reply:
should we allow to override it (via storage.conf) as part of this PR or can it be done later?

Reply:
I've added  a patch that allows setting the minimum and maximum value for the user namespace.

The maximum size is necessary as a malicious image could consume the entire space.